File: dynamic_programming/linear_dp/0053_maximum_subarray.js
/*
Given an integer array nums, find the subarray with the largest sum, 
and return its sum.

A subarray is a contiguous non-empty sequence of elements within an array.

1 <= nums.length <= 10^5
-10^4 <= nums[i] <= 10^4
*/

// https://leetcode.com/problems/maximum-subarray/description/

// brute force -- TLE
// time complexity: O(n^2)
// space complexity: O(1)
var maxSubArray = function (nums) {
  let max = -Infinity;
  for (let i = 0; i < nums.length; i++) {
    let sum = 0;
    for (let j = i; j < nums.length; j++) {
      sum += nums[j];
      max = Math.max(max, sum);
    }
  }
  return max;
};

// dynamic programming
// dp[i] is the maximum subarray sum ending at index i
// dp[i] = max(dp[i-1] + nums[i], nums[i])
// time complexity: O(n)
// space complexity: O(n)
var maxSubArray = function (nums) {
  let dp = new Array(nums.length).fill(0);
  dp[0] = nums[0];
  for (let i = 1; i < nums.length; i++) {
    dp[i] = Math.max(dp[i - 1] + nums[i], nums[i]);
  }
  return Math.max(...dp);
};

// optimized dp
// time complexity: O(n)
// space complexity: O(1)
var maxSubArray = function (nums) {
  for (let i = 1; i < nums.length; i++) {
    nums[i] = Math.max(nums[i], nums[i] + nums[i - 1]);
  }
  return Math.max(...nums);
};

// Kadane's algorithm
// time complexity: O(n)
// space complexity: O(1)
var maxSubArray = function (nums) {
  let sum = 0,
    max = -Infinity;
  for (let i = 0; i < nums.length; i++) {
    sum = Math.max(nums[i], sum + nums[i]);
    max = Math.max(max, sum);
  }
  return max;
};

File: array/prefix_sum/0053_maximum_subarray.py
# Given an integer array nums, find the 
# subarray with the largest sum, and return its sum.

# https://leetcode.com/problems/maximum-subarray/
# similar LC 121 https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

# if preSum < 0: discard
# time complexity O(n)
# space complexity O(1)
class Solution:
    def maxSubArray(self, nums: list[int]) -> int:
        maxSum, preSum = -float('inf'), 0
        for num in nums:
            preSum = max(preSum + num, num) 
            maxSum = max(maxSum, preSum)
        return maxSum
    
# dynamic programming
# let dp[i] denote the max sum of subarray ending at nums[i]
# if dp[i - 1] < 0: dp[i] = nums[i]; if dp[i - 1] >= 0: dp[i] = dp[i - 1] + nums[i]
# thus: dp[i] = max(nums[i], nums[i] + dp[i - 1])
# time complexity O(n)
# space complexity O(n)
class Solution:
    def maxSubArray(self, nums: list[int]) -> int:
        l = len(nums)
        dp = [0 for _ in range(l)]
        dp[0] = nums[0]

        for i in range(1, l):
            dp[i] = max(nums[i], nums[i] + dp[i - 1])
        return max(dp)

# optimized dp: only need max(dp) -- same as preSum approach
# time complexity O(n)
# space complexity O(1)
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        l = len(nums)
        curMax = allMax = nums[0]

        for i in range(1, l):
            curMax = max(curMax + nums[i], nums[i])
            allMax = max(allMax, curMax)
        return allMax
